### PR TITLE
[#4] update branches api documentation

### PR DESCRIPTION
**Pull request by dzakitajudin**

### CODE REVIEW

File: doc/branches.md [hunk 1]
```diff
@@ -1 +1,1845 @@
...
+--- | ---
+Content-Type | application/json
+Accept | application/json
+Email | administrator@gmail.com
+
+### Request Payloads
+Name | Type | Example Value
+--- | --- | ---
+branchcode | string | 12345  
+companycode | string | TES
```
Comment (jacky-htg) [inline doc/branches.md:30]: remove companyCode, you can grab from companyCode of user login

File: doc/branches.md [hunk 2]
```diff
@@ -1 +1,1845 @@
...
+![](./branches-layer.png)
+
+### Headers
+Key | Value 
+--- | ---
+Content-Type | application/json
+Accept | application/json
+Email | administrator@gmail.com
+
+### Request Payloads
```
Comment (jacky-htg) [inline doc/branches.md:26]: use snake case / underscore for two phrase, ex: phone_number

File: doc/branches.md [hunk 3]
```diff
@@ -1 +1,1845 @@
...
+    "faxnumber": "5113470",
+    "email": "testingmotor@gmail.com"
+}
+```
+
+### Response Payloads
+HTTP Code | Status | Description
+--- | --- | ---
+400 | Bad Request | Bad request payload  
+404 | Not Found | User not found in database
```
Comment (jacky-htg) [inline doc/branches.md:62]: remove 404

**Review by jacky-htg — CHANGES_REQUESTED** [review]
Please resolve my comments below. thanks.

File: doc/branches.md [hunk 4]
```diff
@@ -1 +1,1845 @@
 # Branches Module
+Module | HTTP Method | URL | Description 
+--- | --- | --- | ---
+[Create Branches](#create-branches) | POST | /create-branches | Create Branches API
```
[the remark this answers is not in the document]
Comment (dzakitajudin) [reply doc/branches.md]: done

File: doc/branches.md [hunk 5]
```diff
@@ -1 +1,1845 @@
...
+[Edit Branches](#edit-branches) | PUT | /edit-branches/{id} | Edit Branches API
```
[the remark this answers is not in the document]
Comment (dzakitajudin) [reply doc/branches.md]: done

File: doc/branches.md [hunk 6]
```diff
@@ -1 +1,1845 @@
...
+[Delete Branches](#delete-branches) | DELETE | /delete-branches/{id} | Delete Branches API
```
[the remark this answers is not in the document]
Comment (dzakitajudin) [reply doc/branches.md]: done

File: doc/branches.md [hunk 7]
```diff
@@ -1 +1,1845 @@
...
+[Get All Branches](#get-all-branches) | GET | /get-allbranches | Get All Branches API
```
[the remark this answers is not in the document]
Comment (dzakitajudin) [reply doc/branches.md]: done

File: doc/branches.md [hunk 8]
```diff
@@ -1 +1,1845 @@
...
+[Get Branches By ID](#get-branches-byid) | GET | /get-branches-byid/{id} | Get Branches By ID API
```
[the remark this answers is not in the document]
Comment (dzakitajudin) [reply doc/branches.md]: done

File: doc/branches.md [hunk 9]
```diff
@@ -1 +1,1845 @@
...
+[Get Branches By Branch Code](#get-branches-bybranchcode) | GET | /get-branches-bybranchcode/{kode} | Get Branches By Branch Code API
```
[the remark this answers is not in the document]
Comment (dzakitajudin) [reply doc/branches.md]: done

File: doc/branches.md (hunk 1 above)
Comment (dzakitajudin) [reply doc/branches.md]: done

File: doc/branches.md (hunk 2 above)
Comment (dzakitajudin) [reply doc/branches.md:26]: done

File: doc/branches.md (hunk 3 above)
Comment (dzakitajudin) [reply doc/branches.md]: done

File: doc/branches.md [hunk 10]
```diff
@@ -1 +1,1719 @@
...
+### Headers
+Key | Value 
+--- | ---
+Content-Type | application/json
+Accept | application/json
+Email | administrator@gmail.com
+
+### Request Payloads
+No Request Payloads
+
```
Comment (jacky-htg) [inline doc/branches.md:1192]: add param query for filter and pagination like :
- keyword --> for search
- page
- limit
- order
- sort
- filter_field --> if any filter by field

File: doc/branches.md [hunk 11]
```diff
@@ -1 +1,1719 @@
...
+### Response Payloads
+HTTP Code | Status | Description
+--- | --- | ---
+400 | Bad Request | Bad request payload  
+404 | Not Found | User not found in database
```
Comment (jacky-htg) [inline doc/branches.md:1197]: for list, prefer there is no 404. you should use 200 status code with the empty data.

File: doc/branches.md [hunk 12]
```diff
@@ -1 +1,1719 @@
...
+### Response Payloads
+HTTP Code | Status | Description
+--- | --- | ---
+400 | Bad Request | Bad request payload  
+404 | Not Found | User not found in database  
+500 | Internal Server Error | some un-handle error in server 
+200 | OK | OK
+```
+{
+    "status_code": "CDC-404",
```
Comment (jacky-htg) [inline doc/branches.md:1202]: cdc-500

File: doc/branches.md [hunk 13]
```diff
@@ -1 +1,1719 @@
...
+    "status_message": "data not found",
```
Comment (jacky-htg) [inline doc/branches.md:1203]: internal server error

File: doc/branches.md [hunk 14]
```diff
@@ -1 +1,1719 @@
...
+        }
+    ]
+}
+```
+
+### Logic
+
+#### Headers Validation
+- email : user email required and is not empty
+- email : check user email is login or not
```
Comment (jacky-htg) [inline doc/branches.md:1251]: invalid validation

File: doc/branches.md [hunk 15]
```diff
@@ -1 +1,1719 @@
...
+- email : must be top level, like administrator
```
Comment (jacky-htg) [inline doc/branches.md:1252]: use less validation

File: doc/branches.md [hunk 16]
```diff
@@ -1 +1,1719 @@
...
+Response Payload :
+```
+{
+    "status_code": "CDC-400",
+    "status_message": "headers email is required",
+    "data": null
+}
+```
+
+#### Case : Negative Case 2
```
Comment (jacky-htg) [inline doc/branches.md:1274]: invalid case

File: doc/branches.md [hunk 17]
```diff
@@ -1 +1,1719 @@
...
+Response Payload :
+```
+{
+    "status_code": "CDC-400",
+    "status_message": "headers email is not login",
+    "data": null
+}
+```
+
+#### Case : Negative Case 3
```
Comment (jacky-htg) [inline doc/branches.md:1289]: invalid case

File: doc/branches.md [hunk 18]
```diff
@@ -1 +1,1719 @@
...
+    "status_message": "headers email must be top level access or administrator",
+    "data": null
+}
+```
+
+#### Case : Negative Case 4
+
+Data in database is empty
+
+Response HTTP Status Code : 404
```
Comment (jacky-htg) [inline doc/branches.md:1308]: 200 OK

File: doc/branches.md [hunk 19]
```diff
@@ -1 +1,1719 @@
...
+
+Response Payload :
+```
+{
+    "status_code": "CDC-404",
```
Comment (jacky-htg) [inline doc/branches.md:1313]: cdc-200

File: doc/branches.md [hunk 20]
```diff
@@ -1 +1,1719 @@
...
+```
+
+### Logic
+
+#### Endpoint Validation
+must be add parameter id and must exist in database
+
+#### Headers Validation
+- email : user email required and is not empty
+- email : check user email is login or not
```
Comment (jacky-htg) [inline doc/branches.md:1429]: invalid validation

File: doc/branches.md [hunk 21]
```diff
@@ -1 +1,1719 @@
...
+- email : must be top level, like administrator
```
Comment (jacky-htg) [inline doc/branches.md:1430]: use less vaidation

File: doc/branches.md [hunk 22]
```diff
@@ -1 +1,1719 @@
...
+Response Payload :
+```
+{
+    "status_code": "CDC-400",
+    "status_message": "headers email is required",
+    "data": null
+}
+```
+
+#### Case : Negative Case 2
```
Comment (jacky-htg) [inline doc/branches.md:1452]: invalid case

File: doc/branches.md [hunk 23]
```diff
@@ -1 +1,1719 @@
...
+Response Payload :
+```
+{
+    "status_code": "CDC-400",
+    "status_message": "headers email is not login",
+    "data": null
+}
+```
+
+#### Case : Negative Case 3
```
Comment (jacky-htg) [inline doc/branches.md:1467]: invalid case

File: doc/branches.md [hunk 24]
```diff
@@ -1 +1,1719 @@
...
+```
+
+### Logic
+
+#### Endpoint Validation
+must be add parameter branch code and must exist in database
+
+#### Headers Validation
+- email : user email required and is not empty
+- email : check user email is login or not
```
Comment (jacky-htg) [inline doc/branches.md:1596]: invalid validation

File: doc/branches.md [hunk 25]
```diff
@@ -1 +1,1719 @@
...
+- email : must be top level, like administrator
```
Comment (jacky-htg) [inline doc/branches.md:1597]: use less validation

File: doc/branches.md [hunk 26]
```diff
@@ -1 +1,1719 @@
...
+Response Payload :
+```
+{
+    "status_code": "CDC-400",
+    "status_message": "headers email is required",
+    "data": null
+}
+```
+
+#### Case : Negative Case 2
```
Comment (jacky-htg) [inline doc/branches.md:1619]: invalid case

File: doc/branches.md [hunk 27]
```diff
@@ -1 +1,1719 @@
...
+Response Payload :
+```
+{
+    "status_code": "CDC-400",
+    "status_message": "headers email is not login",
+    "data": null
+}
+```
+
+#### Case : Negative Case 3
```
Comment (jacky-htg) [inline doc/branches.md:1634]: invalid case

File: doc/branches.md [hunk 28]
```diff
@@ -1 +1,1719 @@
...
+Response Payload :
+```
+{
+    "status_code": "CDC-400",
+    "status_message": "headers email must be top level access or administrator",
+    "data": null
+}
+```
+
+#### Case : Negative Case 4
```
Comment (jacky-htg) [inline doc/branches.md:1649]: invalid case

**Review by jacky-htg — CHANGES_REQUESTED** [review]

File: doc/branches.md (hunk 10 above)
Comment (jacky-htg) [reply doc/branches.md:866]: what might be considered for adding a warehouse filter?

File: doc/branches.md [hunk 29]
```diff
@@ -1 +1,1266 @@
...
+### Headers
+Key | Value 
+--- | ---
+Content-Type | application/json
+Accept | application/json
+Email | administrator@gmail.com
+
+### Request Param
+GET /branches?keyword="testing"&page=1&limit=20&order=branches.id&sort=desc
+
```
Comment (jacky-htg) [inline doc/branches.md:863]: you can display the query param in table like request payload at create endpoint.

Comment (jacky-htg) [reply doc/branches.md:863]: Name | Example value | Description
--- | --- | ---
keyword | testing | the keyword for searching branches
page | 1 | current page in pagination. default value is 1. if there is no param page, page will be set with 1.

File: doc/branches.md [hunk 30]
```diff
@@ -1 +1,1266 @@
...
+### Headers
+Key | Value 
+--- | ---
+Content-Type | application/json
+Accept | application/json
+Email | administrator@gmail.com
+
+### Request Param
+GET /branches?keyword="testing"&page=1&limit=20&order=branches.id&sort=desc
```
Comment (jacky-htg) [inline doc/branches.md:862]: all param queries must be in the string. so you don't need to quote marks the value.
GET /branches?keyword=testing&page=1&limit=20&order=branches.id&sort=desc&warehouse_id=1,2,3

**Review by jacky-htg — CHANGES_REQUESTED** [review]

File: doc/branches.md (hunk 30 above)
Comment (dzakitajudin) [reply doc/branches.md:862]: i think adding company_id filter is better sir

Comment (jacky-htg) [reply doc/branches.md]: o yeah? who will manage the branches menu? user from cdc? or the main dealer user? 
if the one who manages is the user of cdc, I think it is necessary to add a company_id filter. 
But if the administrator is a user from main dealer, branches are automatically filtered based on the company_id of the currently logged-in user.

**Review by jacky-htg — APPROVED** [review]

**Review by ferrystiawan — APPROVED** [review]

**Review by msyuniarto — APPROVED** [review]

**Pull request merged**